egs/wsj/s5/steps/nnet3/train_discriminative.sh

```diff
@@ -0,0 +1,376 @@
+#!/bin/bash
+
+# Copyright 2012-2014  Johns Hopkins University (Author: Daniel Povey)
+#           2014-2015  Vimal Manohar
+# Apache 2.0.
+
+set -e
+set -o pipefail
+
+# This script does MPE or MMI or state-level minimum bayes risk (sMBR) training
+# using egs obtained by steps/nnet3/get_egs_discriminative.sh
+
+# Begin configuration section.
+cmd=run.pl
+num_epochs=4       # Number of epochs of training;
+                   # the number of iterations is worked out from this.
+                   # Be careful with this: we actually go over the data
+                   # num-epochs * frame-subsampling-factor times, due to
+                   # using different data-shifts.
+use_gpu=true
+truncate_deriv_weights=0  # can be used to set to zero the weights of derivs from frames
+                          # near the edges.  (counts subsampled frames).
+apply_deriv_weights=true
+run_diagnostics=true
+learning_rate=0.00002
+max_param_change=2.0
+scale_max_param_change=false # if this option is used, scale it by num-jobs.
+
+effective_lrate=    # If supplied, overrides the learning rate, which gets set to effective_lrate * num_jobs_nnet.
+acoustic_scale=0.1  # acoustic scale for MMI/MPFE/SMBR training.
+boost=0.0       # option relevant for MMI
+
+criterion=smbr
+drop_frames=false #  option relevant for MMI
+one_silence_class=true # option relevant for MPE/SMBR
+num_jobs_nnet=4    # Number of neural net jobs to run in parallel.  Note: this
+                   # will interact with the learning rates (if you decrease
+                   # this, you'll have to decrease the learning rate, and vice
+                   # versa).
+regularization_opts=
+minibatch_size=64  # This is the number of examples rather than the number of output frames.
+modify_learning_rates=false
+last_layer_factor=1.0  # relates to modify-learning-rates
+first_layer_factor=1.0 # relates to modify-learning-rates
+shuffle_buffer_size=5000 # This "buffer_size" variable controls randomization of the samples
```

Comment:
5k egs is probably too much for discriminative egs as the chunks are large, it will use up too much memory and too much I/O at the start- I'd set it to 1000.